File: tools/python_xtc_filters/xtc_conf.py
#! /usr/bin/env python3
#
# Author: Raul Fernandez Ortega <natambio.audio@gmail.com>, 2022-2026
#
# Licensed under the GNU General Public License v3 (GPLv3); see the LICENSE file.

"""
xtc_conf.py — TOML configuration shared by the two Python XTC generators.

Reads the SAME files as the C tools in tools/xtc_filters, through the standard
library's `tomllib` (Python 3.11+) or `tomli` on older interpreters. The C side
parses a strict subset of TOML by hand; since that subset is a subset, a file
written for either tool is read identically by both.

Schema (see tools/xtc_filters/xtc_conf.h for the C counterpart):

    sample_rate = 48000        # top level, common to the whole design
    filter_len  = 4096

    [xtc]                      # or [left] / [right] in the asymmetric tool
    itd_us      = 170
    ild_db      = 14.0
    ild_alpha   = 2.0
    azimuth_deg = 20

    [output]
    directory = "filters"
    prefix    = "my_room"      # optional

The balance b of the asymmetric model is deliberately absent: it is not baked
into the coefficients but applied downstream as a routing gain. See the balance
section of docs/xtc/xtc_no_simetrico_es.md.
"""

import sys

try:                     # Python 3.11+
    import tomllib
except ModuleNotFoundError:          # pragma: no cover - depends on interpreter
    try:
        import tomli as tomllib
    except ModuleNotFoundError:
        sys.stderr.write(
            "xtc: reading TOML needs Python 3.11+ or the 'tomli' package "
            "(pip install tomli)\n")
        raise

SIDE_KEYS = ("itd_us", "ild_db", "ild_alpha", "azimuth_deg")
SIDE_TYPES = {"itd_us": int, "ild_db": float, "ild_alpha": float, "azimuth_deg": int}

TOP_TYPES = {"sample_rate": int, "filter_len": int}
OUTPUT_KEYS = ("directory", "prefix")


class ConfError(Exception):
    """Configuration problem, already phrased for the user."""


def _load(path):
    try:
        with open(path, "rb") as f:
            return tomllib.load(f)
    except OSError as e:
        raise ConfError("cannot open %s: %s" % (path, e.strerror)) from e
    except tomllib.TOMLDecodeError as e:
        raise ConfError("%s: %s" % (path, e)) from e


def _check_keys(doc, allowed_tables):
    """Rejects unknown keys, the way the C reader does.

    Without this a typo such as `ild_alfa` silently leaves ild_alpha at its
    default, and the parameter appears to do nothing.
    """
    for key, value in doc.items():
        if isinstance(value, dict):
            if key not in allowed_tables:
                raise ConfError("unknown table [%s]" % key)
            known = OUTPUT_KEYS if key == "output" else SIDE_KEYS
            for subkey in value:
                if subkey not in known:
                    raise ConfError("unknown key '%s.%s'" % (key, subkey))
        elif key not in TOP_TYPES:
            raise ConfError("unknown key '%s'" % key)


def _number(value, want, name):
    """Converts with the same strictness as the C side.

    An int is accepted where a float is wanted (TOML has no 14 vs 14.0
    distinction worth enforcing here), but a float where an int is wanted is an
    error: `filter_len = 4096.0` would otherwise truncate silently. Booleans are
    rejected outright — in Python `True` is an int, and `filter_len = true`
    would sail through as 1.
    """
    if isinstance(value, bool):
        raise ConfError("%s must be a number, not a boolean" % name)
    if want is int:
        if not isinstance(value, int):
            raise ConfError("%s must be an integer, got %r" % (name, value))
        return value
    if not isinstance(value, (int, float)):
        raise ConfError("%s must be a number, got %r" % (name, value))
    return float(value)


def _load_side(doc, table, side, required):
    """Overlays doc[table] onto the dict `side`."""
    block = doc.get(table)
    if block is None:
        if required:
            raise ConfError("no [%s] table: an asymmetric design needs both sides"
                            % table)
        return side
    for key in SIDE_KEYS:
        if key in block:
            side[key] = _number(block[key], SIDE_TYPES[key], "%s.%s" % (table, key))
        elif required:
            raise ConfError("[%s] is missing the key '%s'" % (table, key))
    return side


def _load_output(doc, cfg):
    block = doc.get("output")
    if block is None:
        return
    for key in OUTPUT_KEYS:
        if key not in block:
            continue
        value = block[key]
        if not isinstance(value, str):
            raise ConfError("output.%s must be a quoted string" % key)
        if key == "directory" and not value:
            raise ConfError("output.directory must not be empty")
        if key == "prefix" and "/" in value:
            raise ConfError("output.prefix must not contain '/'; "
                            "use output.directory for the path")
        cfg[key] = value


def _load_top(doc, cfg):
    for key, want in TOP_TYPES.items():
        if key in doc:
            cfg[key] = _number(doc[key], want, key)


def load_sym(path, cfg):
    """Overlays `path` onto cfg (a dict), leaving absent keys untouched.

    Every key is optional here, so a stored configuration can be combined with
    command-line flags exactly as in the C tool. Returns cfg.
    """
    doc = _load(path)
    _check_keys(doc, allowed_tables=("xtc", "output"))
    _load_top(doc, cfg)
    _load_side(doc, "xtc", cfg["xtc"], required=False)
    _load_output(doc, cfg)
    return cfg


def load_asym(path, cfg):
    """Reads `path` into cfg (a dict). Returns cfg.

    Both [left] and [right] must be present with all four keys: there is no
    meaningful default for one side of an asymmetric layout, and quietly
    supplying one would produce a plausible filter for a geometry the user never
    described.
    """
    doc = _load(path)
    _check_keys(doc, allowed_tables=("left", "right", "output"))
    _load_top(doc, cfg)
    _load_side(doc, "left", cfg["left"], required=True)
    _load_side(doc, "right", cfg["right"], required=True)
    _load_output(doc, cfg)
    return cfg

File: tools/xtc_filters/xtc_conf.c
/*
 * Author: Raul Fernandez Ortega <natambio.audio@gmail.com>, 2022-2026
 *
 * Licensed under the GNU General Public License v3 (GPLv3); see the LICENSE file.
 */

#include "xtc_conf.h"
#include "toml_min.h"

#include <stdio.h>
#include <string.h>

/* Copies a string value into a fixed buffer, refusing to truncate silently. */
static int copy_str(char *dst, size_t dstlen, const char *src, const char *what,
                    char *err, size_t errlen)
{
    size_t n = strlen(src);
    if (n >= dstlen) {
        snprintf(err, errlen, "%s is too long (%zu characters, maximum %zu)",
                 what, n, dstlen - 1);
        return -1;
    }
    memcpy(dst, src, n + 1);
    return 0;
}

/* Reads the four keys of one side. `required` makes every key mandatory;
 * otherwise absent keys leave *side untouched. Returns 0 on success. */
static int load_side(const toml_min *t, const char *table, xtc_conf_side *side,
                     int required, char *err, size_t errlen)
{
    struct { const char *key; int is_int; void *dst; } fields[] = {
        { "itd_us",      1, &side->itd_us      },
        { "ild_db",      0, &side->ild_db      },
        { "ild_alpha",   0, &side->ild_alpha   },
        { "azimuth_deg", 1, &side->azimuth_deg },
    };

    for (size_t i = 0; i < sizeof(fields) / sizeof(fields[0]); i++) {
        int rc = fields[i].is_int
               ? toml_min_int(t, table, fields[i].key, (int *)fields[i].dst, err, errlen)
               : toml_min_double(t, table, fields[i].key, (double *)fields[i].dst, err, errlen);
        if (rc < 0) return -1;
        if (rc == 0 && required) {
            snprintf(err, errlen, "[%s] is missing the key '%s'", table, fields[i].key);
            return -1;
        }
    }
    return 0;
}

/* Reads [output]; both keys optional. */
static int load_output(const toml_min *t, char *directory, char *prefix,
                       char *err, size_t errlen)
{
    const char *s = NULL;

    int rc = toml_min_string(t, "output", "directory", &s, err, errlen);
    if (rc < 0) return -1;
    if (rc > 0) {
        if (*s == '\0') {
            snprintf(err, errlen, "output.directory must not be empty");
            return -1;
        }
        if (copy_str(directory, XTC_CONF_PATHMAX, s, "output.directory", err, errlen) != 0)
            return -1;
    }

    rc = toml_min_string(t, "output", "prefix", &s, err, errlen);
    if (rc < 0) return -1;
    if (rc > 0) {
        /* A prefix containing a path separator would silently scatter the
         * filters outside output.directory. */
        if (strchr(s, '/') != NULL) {
            snprintf(err, errlen, "output.prefix must not contain '/'; "
                                  "use output.directory for the path");
            return -1;
        }
        if (copy_str(prefix, XTC_CONF_PATHMAX, s, "output.prefix", err, errlen) != 0)
            return -1;
    }
    return 0;
}

/* Reads the two top-level design switches. Both optional: an absent key leaves
 * whatever the caller pre-loaded. Shared by the two loaders because the pair
 * describes the design, not the layout. */
static int load_design(const toml_min *t, int *frac_delay, int *model_delay,
                       char *err, size_t errlen)
{
    if (toml_min_bool(t, "", "frac_delay", frac_delay, err, errlen) < 0) return -1;
    if (toml_min_int(t, "", "model_delay", model_delay, err, errlen) < 0) return -1;
    if (*model_delay < 0) {
        snprintf(err, errlen, "model_delay must not be negative (got %d)", *model_delay);
        return -1;
    }
    return 0;
}

int xtc_conf_load_sym(const char *path, xtc_conf_sym *cfg, char *err, size_t errlen)
{
    static const char *const allowed[] = {
        "sample_rate", "filter_len", "frac_delay", "model_delay",
        "xtc.itd_us", "xtc.ild_db", "xtc.ild_alpha", "xtc.azimuth_deg",
        "output.directory", "output.prefix",
        NULL
    };

    toml_min *t = toml_min_load(path, err, errlen);
    if (!t) return -1;

    if (toml_min_check_keys(t, allowed, err, errlen) != 0) goto fail;

    if (toml_min_int(t, "", "sample_rate", &cfg->sample_rate, err, errlen) < 0) goto fail;
    if (toml_min_int(t, "", "filter_len",  &cfg->filter_len,  err, errlen) < 0) goto fail;

    if (load_design(t, &cfg->frac_delay, &cfg->model_delay, err, errlen) != 0) goto fail;

    if (load_side(t, "xtc", &cfg->xtc, 0, err, errlen) != 0) goto fail;
    if (load_output(t, cfg->directory, cfg->prefix, err, errlen) != 0) goto fail;

    toml_min_free(t);
    return 0;

fail:
    toml_min_free(t);
    return -1;
}

int xtc_conf_load_asym(const char *path, xtc_conf_asym *cfg, char *err, size_t errlen)
{
    static const char *const allowed[] = {
        "sample_rate", "filter_len", "frac_delay", "model_delay",
        "left.itd_us",  "left.ild_db",  "left.ild_alpha",  "left.azimuth_deg",
        "right.itd_us", "right.ild_db", "right.ild_alpha", "right.azimuth_deg",
        "output.directory", "output.prefix",
        NULL
    };

    toml_min *t = toml_min_load(path, err, errlen);
    if (!t) return -1;

    if (toml_min_check_keys(t, allowed, err, errlen) != 0) goto fail;

    if (!toml_min_has_table(t, "left")) {
        snprintf(err, errlen, "no [left] table: an asymmetric design needs both sides");
        goto fail;
    }
    if (!toml_min_has_table(t, "right")) {
        snprintf(err, errlen, "no [right] table: an asymmetric design needs both sides");
        goto fail;
    }

    if (toml_min_int(t, "", "sample_rate", &cfg->sample_rate, err, errlen) < 0) goto fail;
    if (toml_min_int(t, "", "filter_len",  &cfg->filter_len,  err, errlen) < 0) goto fail;

    if (load_design(t, &cfg->frac_delay, &cfg->model_delay, err, errlen) != 0) goto fail;

    if (load_side(t, "left",  &cfg->left,  1, err, errlen) != 0) goto fail;
    if (load_side(t, "right", &cfg->right, 1, err, errlen) != 0) goto fail;
    if (load_output(t, cfg->directory, cfg->prefix, err, errlen) != 0) goto fail;

    toml_min_free(t);
    return 0;

fail:
    toml_min_free(t);
    return -1;
}
